Automattic/wp-calypso
# Browsehappy: More secure redirect

### Changes proposed in this Pull Request
Since we had to remove this redirect feature due to an XSS issue, we needed a more secure way to verify that the redirect URL actually comes from Calypso. I believe this will solve that problem. The goal is to be able to redirect to calypso without being vulnerable to XSS or spoofing.

Here's how it works:
1. When the server redirects to "browsehappy", make sure that "from" starts with the calypso URL. (previously, it would just include the path, like `/posts`. Now it includes the whole URL, like `http://calypso.localhost:3000/posts`.)
2. When the server loads browsehappy, pass in the current calypso URL along with the from query param.
3. Check "from" starts with the current Calypso URL. If so, use "from". If not, set the path to the root.

This means that _our_ version of "from" will include the Calypso URL. If someone else constructs a URL with anything _not_ starting with the Calypso URL, it won't be used.

### Testing instructions
For local testing, comment out lines 33-37 in `middleware/unsupported-browser.js`. Otherwise, I think this should be testable in IE 11 in calypso.live.

1. Load calypso root.
2. You should be redirected to browsehappy 
4. Without clicking "continue", note that the redirect URL is to the calypso root
4. Load "$calypsoUrl/posts" (e.g. calypso.localhost:3000/posts)
5. You should be redirected to browsehappy
4. Without clicking "continue", note that the redirect URL is to /posts
5. Try loading `/browsehappy?from=https://evil.com`. 
6. Without clicking "continue", note that the redirect URL is to the calypso root.
